File: web/tests/settings-presentation.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  appliesText,
  sourceLabel,
  canReset,
  effectiveAsInputValue,
} from "../js/lib/settings-presentation.js";

test("appliesText covers all three real applies values distinctly", () => {
  const immediate = appliesText("immediately");
  const nextSweep = appliesText("next_sweep");
  const restart = appliesText("restart-required");
  assert.notEqual(immediate, nextSweep);
  assert.notEqual(nextSweep, restart);
  assert.notEqual(immediate, restart);
  assert.match(immediate, /immediately/i);
  assert.match(nextSweep, /sweep/i);
  assert.match(restart, /restart/i);
});

test("appliesText falls back honestly for an unknown value rather than lying", () => {
  assert.doesNotThrow(() => appliesText("something-new"));
  assert.notEqual(appliesText("something-new"), appliesText("immediately"));
});

test("sourceLabel covers db/env/default distinctly", () => {
  const labels = new Set([sourceLabel("db"), sourceLabel("env"), sourceLabel("default")]);
  assert.equal(labels.size, 3);
});

test("canReset: only a db-sourced, non-env-only entry can be reset", () => {
  assert.equal(canReset({ source: "db", env_only: false }), true);
  assert.equal(canReset({ source: "env", env_only: false }), false);
  assert.equal(canReset({ source: "default", env_only: false }), false);
  assert.equal(canReset({ source: "db", env_only: true }), false);
});

test("effectiveAsInputValue: null becomes blank (the disabled state, not the string 'null')", () => {
  assert.equal(effectiveAsInputValue({ key: "schedule_window", effective: null }), "");
});

test("effectiveAsInputValue: a list becomes a comma-joined string", () => {
  assert.equal(
    effectiveAsInputValue({ key: "webhook_events", effective: ["job.done", "job.error"] }),
    "job.done,job.error",
  );
});

test("effectiveAsInputValue: numbers and plain strings pass through as strings", () => {
  assert.equal(effectiveAsInputValue({ key: "schedule_interval_minutes", effective: 180 }), "180");
  assert.equal(effectiveAsInputValue({ key: "vault_name", effective: "vault-01" }), "vault-01");
});

test("effectiveAsInputValue: empty list becomes empty string, not '[]' or a stray comma", () => {
  assert.equal(effectiveAsInputValue({ key: "webhook_events", effective: [] }), "");
});

File: app/app/src/test/java/dev/steamvault/app/ui/settings/logic/SettingsPresentationTest.kt
package dev.steamvault.app.ui.settings.logic

import dev.steamvault.app.net.model.SettingInfoOut
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class SettingsPresentationTest {

    private fun entry(
        key: String = "vault_name",
        effective: kotlinx.serialization.json.JsonElement = JsonPrimitive("vault-01"),
        source: String = "db",
        fallback: kotlinx.serialization.json.JsonElement = JsonNull,
        applies: String = "immediately",
        envOnly: Boolean = false,
    ) = SettingInfoOut(key, effective, source, fallback, applies, envOnly)

    // ---- parseSettingsApplies -----------------------------------------------

    @Test
    fun `parses all three known applies values`() {
        assertEquals(SettingsApplies.IMMEDIATELY, parseSettingsApplies("immediately"))
        assertEquals(SettingsApplies.NEXT_SWEEP, parseSettingsApplies("next_sweep"))
        assertEquals(SettingsApplies.RESTART_REQUIRED, parseSettingsApplies("restart-required"))
    }

    @Test
    fun `MUTATION PIN -- an unrecognised applies value falls back to UNSPECIFIED, not a silent default`() {
        assertEquals(SettingsApplies.UNSPECIFIED, parseSettingsApplies("whenever-i-feel-like-it"))
    }

    // ---- parseSettingsSource ------------------------------------------------

    @Test
    fun `parses all three known source values`() {
        assertEquals(SettingsSource.DB, parseSettingsSource("db"))
        assertEquals(SettingsSource.ENV, parseSettingsSource("env"))
        assertEquals(SettingsSource.DEFAULT, parseSettingsSource("default"))
    }

    @Test
    fun `MUTATION PIN -- an unrecognised source value falls back to UNKNOWN`() {
        assertEquals(SettingsSource.UNKNOWN, parseSettingsSource("mystery"))
    }

    // ---- canResetSetting ------------------------------------------------------

    @Test
    fun `a db-sourced, non-env-only entry can be reset`() {
        assertTrue(canResetSetting(entry(source = "db", envOnly = false)))
    }

    @Test
    fun `MUTATION PIN -- an env-sourced entry cannot be reset even if source string is db-like elsewhere`() {
        assertFalse(canResetSetting(entry(source = "env")))
        assertFalse(canResetSetting(entry(source = "default")))
    }

    @Test
    fun `MUTATION PIN -- an env-only entry can never be reset, even if it somehow reports source db`() {
        assertFalse(canResetSetting(entry(source = "db", envOnly = true)))
    }

    // ---- effectiveAsFieldText -------------------------------------------------

    @Test
    fun `a string effective value round-trips as-is`() {
        assertEquals("vault-01", effectiveAsFieldText(entry(effective = JsonPrimitive("vault-01"))))
    }

    @Test
    fun `an int effective value becomes its decimal text`() {
        assertEquals("30", effectiveAsFieldText(entry(effective = JsonPrimitive(30))))
    }

    @Test
    fun `a JSON null effective value becomes an empty string`() {
        assertEquals("", effectiveAsFieldText(entry(effective = JsonNull)))
    }

    @Test
    fun `a webhook_events array becomes a comma-joined string`() {
        val effective = JsonArray(listOf(JsonPrimitive("job.done"), JsonPrimitive("job.error")))
        assertEquals("job.done,job.error", effectiveAsFieldText(entry(key = "webhook_events", effective = effective)))
    }

    @Test
    fun `an empty webhook_events array becomes an empty string, not a stray comma`() {
        assertEquals("", effectiveAsFieldText(entry(key = "webhook_events", effective = JsonArray(emptyList()))))
    }
}
